enhancements/dns/cache-tuning.md

```diff
@@ -0,0 +1,245 @@
+A. Yes. CoreDNS uses `time.Duration` for TTLs which enables using extremely long durations.
+
+Q. Will CoreDNS transparently pass through TTLs from upstream resolvers?
+A.
+
+Q. Should this be a top level API that applies to all server blocks or should this be configurable per server?
+A.
+
+### Test Plan
+
+#### Testing TTL settings
+1. Deploy an upstream resolver.
+2. Modify the cache settings to have a low `successTTL` (e.g. `1m`).
+3. Issue a known successful DNS query for the associated zone once to ensure the record is cached.
+4. Issue the query a second time to ensure that it's served from the cache.
+5. Wait until the TTL period has expired and then issue the query again to ensure that it's not served from the cache.
+
+#### Testing TTL Passthrough
+1. Deploy an upstream resolver that can respond with a specific TTL for a zone. Ensure that this TTL is higher than 
+   the default for CoreDNS.
+2. Issue a known successful query for the associated zone.
+3. Wait until the cluster DNS TTL has expired and then issue the query again to ensure that it's survived the 
+   cluster DNS cache settings.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+This will go directly to GA.
+
+#### Tech Preview -> GA
+
+This will go directly to GA. It will include product documentation changes and e2e tests.
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+Upgrade expectations:
+- When upgrading to versions where the cache API _is not_ available, the current default in the CoreDNS Corefile will 
+  apply normally.
+- When upgrading to versions where the cache API _is_ available, no interruption will occur as the CoreDNS Corefile 
+  cache settings will apply by default even if the API is not configured.
+- When upgrading from a version with the cache API to another version with the cache API, the configuration will be 
+  retained across upgrades.
+
+Downgrade expectations:
+- When downgrading to versions where the cache API _is not_ available, the cluster admin will remove the caching 
+  settings when downgrading.
+- When downgrading to versions where the cache API _is_ available, the configuration settings should be retained 
+  across downgrades.
+
+### Version Skew Strategy
+
+N/A
+
+### Operational Aspects of API Extensions
+
+Since allowing for adjustment of the CoreDNS cache settings is primarily a judgement call based on the environment 
+the cluster is running in, we do not expect a change in operator conditions as a result of configuring caching. 
+Modifying cache settings could have an effect on the collected CoreDNS metrics and as a result the alerts associated 
+with those metrics but they would be localized to a particular environment and would need to be triaged accordingly.
+
+#### Failure Modes
+
+* TTLs set too low
+  * This could cause cache churn and put excessive load on CoreDNS. 
+* TTLs set too high
+  * This could cause stale data to be served back to clients.
+
```

Comment:
That's a good question. I'll post the 4.10 Corefile and my rationale in case there are any holes in it.

4.10 Corefile:
```
    .:5353 {
        bufsize 512
        errors
        log . {
            class error
        }
        health {
            lameduck 20s
        }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
            pods insecure
            fallthrough in-addr.arpa ip6.arpa
        }
        prometheus 127.0.0.1:9153
        forward . /etc/resolv.conf {
            policy sequential
        }
        cache 900 {
            denial 9984 30
        }
        reload
    }
```
The `kubernetes` plugin is authoritative for `cluster.local`, `in-addr.arpa`, and `ip6.arpa` in that configuration. Since `cluster.local` is not in the fallthrough, if it can't be resolved then I believe processing will stop. If it can be resolved, then it gets passed to the remaining plugins so you could conceivably affect critical cluster services. @Miciah can keep me honest here but I believe that's how it'd function.

Is this an area where it makes sense to have a minimum given that the use case for this feature is generally to increase the caching to reduce load?

Reply:
Should we cache anything from the `kubernetes` plugin at all?

I realise that would be a breaking change but is the ask to configure cache behaviour for the `forward` plugin only? If so, could we remove caching from the `kubernetes` plugin by having two server blocks:

```
cluster.local:53 {
    bufsize 512
    errors
    debug
    log
    ready
    kubernetes cluster.local in-addr.arpa ip6.arpa {
        kubeconfig /home/aim/.kube/config
        pods insecure
        fallthrough in-addr.arpa ip6.arpa
    }
}

.:53 {
    bufsize 512
    errors
    log . {
      class success
      class denial
    }
    health {
        lameduck 1s
    }
    ready
    forward . /etc/resolv.conf {
        policy sequential
    }
    cache 900 {
        denial 9984 30
    }
    reload
}

```

Removing any caching from cluster local results might address some of @jerpeter1 concerns.

Reply:
I don't know whether the cache plugin caches responses from the kubernetes plugin.  That would be good to investigate (one way would be to monitor the cache hit metric while doing some service lookups).  I don't think that there's any issue here though; the new configuration option only affects the max allowed TTL, so it's up to the kubernetes plugin to set a reasonable TTL on the responses that it synthesizes.

Reply:
I've confirmed that the `cache` plugin caches the responses from the `kubernetes` plugin using @Miciah's method. On a stable cluster, I ran a service lookup in a `while true` loop and the cache hit count went up sharply on the metrics graph.

I pushed up some changes to this effect but I found after pushing them up that `kubernetes` and `cache` both have a default minimum TTL of `5s` so I need to run another test to disambiguate whether the `5s` TTL in a query response for a cluster service is coming from `kubernetes` or `cache`

Reply:
So it does appear that the TTL that comes back in the response is coming from the `cache` plugin.

Corefile:
```
.:5353 {
    bufsize 512
    errors
    log . {
        class error
    }
    health {
        lameduck 20s
    }
    ready
    kubernetes cluster.local in-addr.arpa ip6.arpa {
        pods insecure
        fallthrough in-addr.arpa ip6.arpa
    }
    prometheus 127.0.0.1:9153
    forward . /etc/resolv.conf {
        policy sequential
    }
    cache {
        success 9984 900 10
        denial 9984 30 10
    }
    reload
}
```

Query response to a cluster service.
```shell
sh-4.4# dig +nocmd +noquestion +nostats +nocomments +ttlunits redhat-marketplace.svc.cluster.local @172.30.0.10                                                                                                   
cluster.local.          10s     IN      SOA     ns.dns.cluster.local. hostmaster.cluster.local. 1659121587 7200 1800 86400 5
```

Reply:
Updated the EP with some test scenarios. We'll also cover this nuance with e2e tests as well.